src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Hover/HoverFactory.cs

```diff
@@ -82,20 +82,20 @@ internal static class HoverFactory
             if (binding is null)
             {
                 // No matching tagHelpers, it's just HTML
-                return SpecializedTasks.Null<VSInternalHover>();
+                return SpecializedTasks.Null<LspHover>();
             }
             else if (binding.IsAttributeMatch)
             {
                 // Hovered over a HTML tag name but the binding matches an attribute
-                return SpecializedTasks.Null<VSInternalHover>();
+                return SpecializedTasks.Null<LspHover>();
             }
 
             Debug.Assert(binding.Descriptors.Any());
 
             var span = containingTagNameToken.GetLinePositionSpan(codeDocument.Source);
 
             return ElementInfoToHoverAsync(
-                documentFilePath, binding.Descriptors, span, options, solutionQueryOperations, cancellationToken);
+                codeDocument.Source.FilePath, binding.Descriptors, span, options, solutionQueryOperations, cancellationToken);
```

Comment:
Warning: Thoughts ahead!

I don't think I realised that a file path was attainable from a `RazorCodeDocument`, and now I'm wondering if there are other places where I've pass in an `IDocumentSnapshot` that could actually have taken a `RazorCodeDocument`.

But I'm also wondering the opposite. Presumably this `codeDocument` came from an `IDocumentSnapshot` in the caller, so would it make sense to pass an `IDocumentSnapshot` all the way through to the `GetProjectAvailability` methods? In LSP it wouldn't change much, but perhaps it would allow for smarter code in Roslyn, is there if an existing way to go from a `TextDocument` to all of the projects it is contained in. Though I don't think that fits with the current solution query API, so probably unnecessarily overcomplicated for now.

Reply:
There comes a point where it bumps into completion needing tooltips as well. So, at most we would have only been able to push the `RazorCodeDocument` one method call further. Once everything's happily in co-hosting, I think that's a good change to make.